### PR TITLE
Add project-filter field to App Engine deploy dialog

### LINKED ISSUE #1380: Search field to filter projects

### PR DESCRIPTION
**Pull request by briandealwis**
Also did some minor refactoring of the `ProjectSelector` to hide implementation details.

screen shot 2017-07-11 at 2 28 14 pm

Fixes #1380

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/AppEngineDeployPreferencesPanel.java [hunk 1]
```diff
@@ -375,6 +384,14 @@ public void widgetSelected(SelectionEvent event) {
         new ProjectSelectorSelectionChangedListener(accountSelector,
                                                     projectRepository,
                                                     projectSelector));
+    filterField.addModifyListener(new ModifyListener() {
+      @Override
+      public void modifyText(ModifyEvent e) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/AppEngineDeployPreferencesPanel.java]: e --> event

File: plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/ProjectSelector.java [hunk 2]
```diff
@@ -74,36 +80,85 @@ public ProjectSelector(Composite parent) {
...
+   */
+  public void setFilter(final String searchText) {
+    if (searchText == null || Strings.isNullOrEmpty(searchText)) {
+      viewer.resetFilters();
+      return;
+    }
+    ViewerFilter filter = new ViewerFilter() {
+      @Override
+      public boolean select(Viewer viewer, Object parentElement, Object element) {
+        for (IValueProperty prop : projectProperties) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/ProjectSelector.java]: prop --> property

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/AppEngineDeployPreferencesPanel.java [hunk 3]
```diff
@@ -354,6 +356,13 @@ private void createProjectIdSection() {
     GridLayoutFactory.fillDefaults().numColumns(2).spacing(0, 0).applyTo(projectSelectorComposite);
     GridDataFactory.fillDefaults().grab(true, false).applyTo(projectSelectorComposite);
 
+    final Text filterField =
+        new Text(projectSelectorComposite,
+            SWT.BORDER | SWT.SEARCH | SWT.ICON_SEARCH | SWT.ICON_CANCEL);
+    filterField.setMessage("Filter projects");
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/AppEngineDeployPreferencesPanel.java]: 1. This should be in the message bundle.
2. by name, by ID, or both?

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/ConvertJobSuspender.java [hunk 4]
```diff
@@ -61,7 +61,7 @@ private SuspendedJob(Job job, long scheduleDelay) {
   static Predicate<Job> isConvertJob = new Predicate<Job>() {
     @Override
     public boolean apply(Job job) {
-      return CONVERT_JOB_CLASS_NAME.equals(job.getClass().getName());
+      return job != null && CONVERT_JOB_CLASS_NAME.equals(job.getClass().getName());
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/ConvertJobSuspender.java]: Was there a problem with this? This makes me think `job` can actually be null, but I don't see how that's possible.

File: plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/ProjectSelector.java [hunk 5]
```diff
@@ -74,36 +81,101 @@ public ProjectSelector(Composite parent) {
...
+
+  /**
+   * Set a search filter on the list. If empty or {@code null}, then removes any existing filters.
+   */
+  public void setFilter(final String searchText) {
+    if (searchText == null || Strings.isNullOrEmpty(searchText)) {
+      viewer.resetFilters();
+      return;
+    }
+    final String[] searchTerms = searchText.split("\\s");
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/ProjectSelector.java]: Could be put into `select() below`.

File: plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/ConvertJobSuspender.java (hunk 4 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/ConvertJobSuspender.java]: Predicate doesn't declare its argument as `@NonNull`.  I'll drop this out of this PR though.

File: plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/ProjectSelector.java [hunk 6]
```diff
@@ -74,36 +80,85 @@ public ProjectSelector(Composite parent) {
...
     }
-    tableViewer.setSelection(selection);
+    viewer.setSelection(selection);
   }
 
+  /**
+   * Set a search filter on the list. If empty or {@code null}, then removes any existing filters.
+   */
+  public void setFilter(final String searchText) {
+    if (searchText == null || Strings.isNullOrEmpty(searchText)) {
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.projectselector/src/com/google/cloud/tools/eclipse/projectselector/ProjectSelector.java]: Good point.

**Pull request merged**